File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys
input = sys.stdin.readline

n, c = map(int, input().split())
home = list(int(input()) for _ in range(n))

home.sort()

def get_wifi_num(gap):
    pre = home[0]
    count = 1
    for cur in home:
        if cur - pre >= gap:
            pre = cur
            count += 1
    return count

answer = 0
left = 1
right = home[-1] - home[0]
while left <= right:
    mid = (left + right) // 2
    wifi = get_wifi_num(mid)
    
    if wifi >= c:
        left = mid + 1
        answer = mid
    else:
        right = mid - 1

print(answer)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().trim().split('\n')

const [N, C] = input.shift().split(' ').map(Number)
const wifi = input.map(Number)
wifi.sort((a, b) => a - b)

let answer = 1
let left = 1
let right = Math.max(...wifi)
while(left <= right) {
  const mid = Math.floor((left + right) / 2)
  const num = add_wifi(mid)

  if (C <= num) {
    answer = Math.max(answer, mid)
    left = mid + 1
  } else {
    right = mid - 1
  }
}
console.log(answer)

function add_wifi(minGap) {
  let cnt = 1
  let pre = wifi[0]

  for (let i = 1; i < N; i++) {
    const gap = wifi[i] - pre

    if (gap < minGap) continue
    pre = wifi[i]
    cnt++
  }

  return cnt
}